apps/heft/src/cli/HeftCommandLineUtilities.ts

```diff
@@ -249,60 +139,93 @@ export class HeftCommandLineUtilities {
   ): IHeftStringListParameter {
-    const actionParameterMap: Map<CommandLineAction, CommandLineStringListParameter> = new Map();
-    for (const action of this._getActions(options.associatedActionNames)) {
+    return this._registerParameter(
+      options,
+      (action: CommandLineAction) => action.defineStringListParameter(options),
+      (parameter: CommandLineStringListParameter) => parameter.values,
+      (parameter: CommandLineStringListParameter) => (parameter.values || []).length > 0
+    );
+  }
+
+  private _registerParameter<TTSCommandLineParameter extends CommandLineParameter, TValue>(
+    options: IRegisterParameterOptions,
+    defineParameterForAction: (action: CommandLineAction) => TTSCommandLineParameter,
+    getParameterValue: (parameter: TTSCommandLineParameter) => TValue | undefined,
+    getValueIsProvided: (parameter: TTSCommandLineParameter) => boolean
+  ): IHeftBaseParameter<TValue> {
+    const actionParameterMap: Map<CommandLineAction, TTSCommandLineParameter> = new Map();
+    for (const action of this._getActions(options.associatedActionNames, options.parameterLongName)) {
       this._verifyUniqueParameterName(action, options);
-      const parameter: CommandLineStringListParameter = action.defineStringListParameter(options);
+      const parameter: TTSCommandLineParameter = defineParameterForAction(action);
       actionParameterMap.set(action, parameter);
     }
-    const parameterObject: Partial<IHeftStringListParameter> = {};
-    Object.defineProperties(parameterObject, {
-      values: {
-        get: (): readonly string[] | undefined => {
-          this._verifyParametersProcessed(options.parameterLongName);
-          if (this._commandLineParser.selectedAction) {
-            return actionParameterMap.get(this._commandLineParser.selectedAction)?.values;
+
+    const parameterObject: IHeftBaseParameter<TValue> = Object.defineProperties(
+      {} as IHeftBaseParameter<TValue>,
+      {
+        value: {
+          get: (): TValue | undefined => {
+            this._verifyParametersProcessed(options.parameterLongName);
+            if (this._commandLineParser.selectedAction) {
+              const parameter: TTSCommandLineParameter | undefined = actionParameterMap.get(
+                this._commandLineParser.selectedAction
+              );
+              if (parameter) {
+                return getParameterValue(parameter);
+              }
+            }
+
+            return undefined;
           }
-        }
-      },
-      actionAssociated: {
-        get: (): boolean => {
-          if (this._commandLineParser.selectedAction) {
-            if (actionParameterMap.get(this._commandLineParser.selectedAction)) {
-              return true;
+        },
+
+        actionAssociated: {
+          get: (): boolean => {
+            if (this._commandLineParser.selectedAction) {
+              if (actionParameterMap.get(this._commandLineParser.selectedAction)) {
+                return true;
```

Comment:
@iclanton  It seems like `actionAssociated` is going to start out as initially `false` and then magically become `true` at some point during the plugin lifecycle, when parsing has occurred.  In order to document this clearly, we need some Heft event name that indicates when the value is assigned.  (It's also tempting to throw an exception or return `undefined` or something other than `false` when the value is unknown, but ideally the lifecycle would be dead simple enough that we can keep this as a  simple boolean.)

Reply:
Done.